model/DataBaseAccess.php

```diff
@@ -65,6 +68,283 @@ protected function getEventManager()
+
+        foreach ($results as $result) {
+            $data[$result[self::COLUMN_RESOURCE_ID]][$result[self::COLUMN_USER_ID]][] = $result[self::COLUMN_PRIVILEGE];
+        }
+
+        return $data;
+    }
+
+    public function addResourcesPermissions(array $resources): void
+    {
+        $insert = [];
+
+        foreach ($resources as $resource) {
+            foreach ($resource['permissions']['add'] as $userId => $addPermissions) {
+                if (empty($addPermissions)) {
+                    continue;
+                }
+
+                foreach ($addPermissions as $permission) {
+                    $insert[] = [
+                        self::COLUMN_USER_ID => $userId,
+                        self::COLUMN_RESOURCE_ID => $resource['id'],
+                        self::COLUMN_PRIVILEGE => $permission,
+                    ];
+                }
+            }
+        }
+
+        $this->insertPermissions($insert);
+
+        foreach ($insert as $inserted) {
+            $this->getEventManager()->trigger(
+                new DacAddedEvent(
+                    $inserted[self::COLUMN_USER_ID],
+                    $inserted[self::COLUMN_RESOURCE_ID],
+                    (array) $inserted[self::COLUMN_PRIVILEGE]
+                )
+            );
+        }
+    }
+
+    public function removeResourcesPermissions(array $resources): void
+    {
+        $groupedRemove = [];
+        $eventsData = [];
+
+        foreach ($resources as $resource) {
+            foreach ($resource['permissions']['remove'] as $userId => $removePermissions) {
+                if (empty($removePermissions)) {
+                    continue;
+                }
+
+                $idString = implode($removePermissions);
+
+                $groupedRemove[$userId][$idString]['resources'][] = $resource['id'];
+                $groupedRemove[$userId][$idString]['privileges'] = $removePermissions;
+
+                $eventsData[] = [
+                    'userId' => $userId,
+                    'resourceId' => $resource['id'],
+                    'privileges' => $removePermissions,
+                ];
+            }
+        }
+
+        foreach ($groupedRemove as $userId => $priviligeGroups) {
+            foreach ($priviligeGroups as $privilegeGroup) {
+                $inQueryPrivilege = implode(',', array_fill(0, count($privilegeGroup['privileges']), ' ? '));
+                $inQueryResources = implode(',', array_fill(0, count($privilegeGroup['resources']), ' ? '));
+
+                $query = <<<SQL
+DELETE
+FROM data_privileges
+WHERE resource_id IN ($inQueryResources)
+  AND privilege IN ($inQueryPrivilege)
+  AND user_id = ?
+SQL;
+
+                $params = array_merge(
+                    array_values($privilegeGroup['resources']),
+                    array_values($privilegeGroup['privileges']),
+                    [$userId]
+                );
+
+                $this->getPersistence()->exec($query, $params);
+            }
+        }
```

Comment:
Remember to wrap this in a transaction (the same way we do for INSERTs [here](https://github.com/oat-sa/extension-tao-dac-simple/blob/9839ed7f372a9c640d37adaca2ee33140b6b53f1/model/DataBaseAccess.php#L738C9-L738C36)).